### PR TITLE
Actually use the IR implementation of hijacked classes in the Emitter.

### PR DESCRIPTION
**Pull request by sjrd**
The main topic of this PR is that the `Emitter` now actually uses the IR of hijacked classes. Instead of duplicating the implementation of all the methods of hijacked classes in `scalajsenv.js`, they are now generated by the Emitter from the IR, in a way very similar to default methods.

Once we do this, we can finally fix the IR of `String` to actually contain all its methods, rather than having most of them in `RuntimeString`, with compiler hacks to reroute them that way.

### CODE REVIEW

**Comment by sjrd** [comment]
Ready for review. @gzm0 I think you'll like this one :)

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -2336,44 +2379,28 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
...
+      case ByteType       => Definitions.BoxedByteClass
+      case ShortType      => Definitions.BoxedShortClass
+      case IntType        => Definitions.BoxedIntegerClass
+      case LongType       => Definitions.BoxedLongClass
+      case FloatType      => Definitions.BoxedFloatClass
+      case DoubleType     => Definitions.BoxedDoubleClass
+      case StringType     => Definitions.StringClass
+    }
+
+    /* Ideally, we should figure out this set using a CHA. However, this is
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala]: CHA?

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 2]
```diff
@@ -3811,28 +3805,16 @@ abstract class GenJSCode extends plugins.PluginComponent
...
-                      methodIdent,
-                      castCallTrg :: arguments,
-                      toIRType(retTpe)),
-                  retTpe)
-            }
+            val retTpe = methodInStringClass.tpe.resultType
+            val castCallTrg = fromAny(callTrg, StringClass.toTypeConstructor)
+            boxIfNeeded(
+                genApplyMethod(castCallTrg, methodInStringClass, arguments),
+                retTpe)
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Why is string any different here than any other hijacked class?

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 3]
```diff
@@ -5451,13 +5428,18 @@ abstract class GenJSCode extends plugins.PluginComponent
   private def isStringType(tpe: Type): Boolean =
     tpe.typeSymbol == StringClass
 
-  protected lazy val isHijackedBoxedClass: Set[Symbol] = {
-    /* This list is a duplicate of ir.Definitions.HijackedBoxedClasses, but
+  protected lazy val isHijackedClass: Set[Symbol] = {
+    /* This list is a duplicate of ir.Definitions.HijackedClasses, but
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Do we still need `ir.Definitions.HijackedBoxedClasses`? If yes, why?

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala (hunk 1 above)
Comment (sjrd) [reply tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/FunctionEmitter.scala]: Class Hierarchy Analysis. I thought that was a sufficiently well-known term. I'll elaborate.

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala (hunk 3 above)
Comment (sjrd) [reply compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Right now, there are still two cases in the Emitter that use `ir.Definitions.HijackedBoxedClasses`. However that's limited to how the emitter does its job, and that definition should definitely be moved over there. However, I suggest we don't do that in this PR as it is not directly related. There are other things in `Definitions` that should be removed, btw.

I've filed #3195.

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala (hunk 2 above)
Comment (sjrd) [reply compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: It isn't (apart from the fact that other hijacked classes also have an associated primitive type). There is indeed some factoring out to do here.

**Comment by sjrd** [comment]
Updated.

**Pull request merged**